docs2/site/docs/guides/dataloader.md

```diff
@@ -73,7 +73,10 @@ First, inject the `IDataLoaderContextAccessor` into your GraphQL type class.
 
 Then use the `Context` property on the accessor to get the current `DataLoaderContext`. The `DataLoaderDocumentListener` configured above ensures that each request will have its own context instance.
 
-Use one of the "GetOrAddLoader" methods on the `DataLoaderContext`. These methods all require a string key to uniquely identify each loader. They also require a delegate for fetching the data. Each method will get an existing loader or add a new one, identified by the string key. Each method has various overloads to support different ways to load and map data with the keys.
+Use one of the `GetOrAdd*Loader` methods on the `DataLoaderContext`. These methods all require a string key to uniquely identify each loader. They also require a delegate for fetching the data. Each method will get an existing loader or add a new one, identified by the string key. Each method has various overloads to support different ways to load and map data with the keys. The main difference you will see is between the `GetOrAddBatchLoader` and `GetOrAddCollectionBatchLoader` methods:
+
+- use `GetOrAddBatchLoader` for one-to-one relationships, i.e. when each key links to a single value; think for example of resolving users by their ids: each user id maps to one (and only one) user;
+- use `GetOrAddCollectionBatchLoader` for one-to-many or many-to-one relationships, i.e. when each key links to multiple values; think for example of resolving orders by user id: each user id may map to (zero or) more orders.
```

Comment:
I think the terms 'one-to-one relationship' and 'one-to-many or many-to-one relationships' may be a bit misleading here.  If we consider typical database diagrams, such as Products and Categories, where each product maps to exactly one category, this is a typical many-to-one relationship yet it is `GetOrAddBatchLoader` which would be applicable for the field named `category` from within the Product type.   And `GetOrAddCollectionBatchLoader` may be used for one-to-many relationships (e.g. 'products' field under the Category type), or many-to-many relationships (like if each product could link to multiple categories).

One-to-one relationships could certainly be represented by `GetOrAddBatchLoader`, or it is likely enough that they are an implementation detail, not exposed through the GraphQL schema.  I honestly have not used one-to-one relationships in any of my databases so I am not sure how to formulate an example.

I suggest also that the example below be expanded, perhaps with three fields (many-to-one, one-to-many, and many-to-many) with applicable comments so people can better visualize the differences.

Here's an example schema, but perhaps you can think of something better:

```gql
type Product {
  # other fields
  brand: Brand!             # Many-to-One relationship: Products belong to one Brand
  categories: [Category!]!  # Many-to-Many relationship: Products belong to many Categories
  reviews: [Review!]!       # One-to-Many relationship: A Product can have multiple Reviews
}
```

Reply:
I see what you mean, but I
I think that it would it make it less confusing if we just focussed on the use case for the graphql API rather than trying to map this back to some relational theory. After all, GraphQL is not tied to any particular storage. The way I see it, the N+1 problem arises when you have a query which involves a list of entities, which you also want to resolve some field of, regardless of whether that field is a list or a single type. In the example below:

- `customers` will resolve to a list. For each customer, 
   - `defaultAddress` will resolve to a single entity.
   - `orders`  will resolve to a list. For each order
       - `products` will resolve to a list.

```graphql
query NPlusOne {
    customers {
        # one-to-one, one customer -> one address
        defaultAddress {
             street
             number
             postCode
        }
        # one-to-many, one customer -> many orders
        orders {
            # many-to-many, one order can link to multiple products and one product can be in multiple orders
            products { 
              qty
              price
              total
            }
        }
    }
}
```

When resolving a field like `customers.defaultAddress`, we run into the N+1 problem as GraphQL tries to resolve each address separately by looping over the items in the `customers` list. A data loader simply informs the resolution strategy to wait until all customers are known before resolving their address via some special resolver.

Since each customer can have (in our simplified model) just one address, we would use `GetOrAddBatchLoader` and have some resolver in the `CustomerType` class like

```csharp
Field<DefaultAddressType, DefaultAddress>()
    .Name("defaultAddress")
    .ResolveAsync(context =>
    {
        var loader = accessor.Context.GetOrAddBatchLoader<int, DefaultAddress>("GetDefaultAddressForCustomers", customers.GetDefaultAddress);
    
        // Add this UserId to the pending keys to fetch
        return loader.LoadAsync(context.Source.CustomerId);
    });
```

For `customers.orders` the problem is the same, but this time each customer can have multiple orders on their name. The same goes for `orders.products`. So in both cases we would use `GetOrAddCollectionBatchLoader` instead

Reply:
I've updated the docs with some of the above reasoning. Hopefully it is now more clear?

Reply:
> it would make it less confusing if we just focussed on the use case for the graphql API rather than trying to map this back to some relational theory

Well, GraphQL is designed specifically for relational data.  That's what the "graph" in GraphQL is; each type is linked to other types (relationships) and the protocol allows for requesting data across those defined relationships.  While it's possible to use GraphQL to communicate with a non-relational backend, such as a simple IOT device or data logger, a simple REST API would suffice.

And specifically here the topic is data loaders, which explicitly relevant to relationships.  As you stated, we should "focus on the use case for the GraphQL API" which in this case most decidedly is "relational theory".

On a side note, an address for a customer is likely a one-to-one relationship - at least from the point of view of the defined relationships within GraphQL.  Likely enough on the database side it's a many-to-one relationship, so the address entity can be linked to multiple orders without duplication, but assuming that there is no `orders` field on the `Address` type, it would be perceived by the GraphQL client as a one-to-one relationship.